### DISCUSSION TITLE
How to set stack size limit?

### DISCUSSION THREAD

**Discussion by jcubic** [nodejs/help#5014] [Q&A]
It seems that `node --stack-size` doesn't work anymore with Node 18. I can't find it when calling `node --help` or `man node`.

I need this because V8 or NodeJS randomly increased Stack size and it breaks my tests. Because I test if there is a stack overflow error when testing TCO (tail call optimization) that is not implemented yet. I need stack overflow. I can increase the number of interactions but then Node can change the limit again randomly. Since computers become more powerful with more ram.

I'm testing this code:

```
node --stack-size=256 $(npm bin)/c8 ava --timeout 10m --verbose tests/test.js
```
Which runs C8 and Ava to test my code. But it seems that the node option make no effect.

**Comment by bnoordhuis** [comment]
It sounds like you think `--stack-size=<n>` sets the stack size but in fact it instructs node to assume the stack is _this_ big. Set it to a value that's below what `ulimit -s` prints.

**Reply by jcubic** [reply]
This is what `ulimit` returns:
```
$ ulimit -s
8192
```

256 is smaller than 8192. It seems that it makes no effect and throws an error in my code that gives stack overflow.

**Reply by bnoordhuis** [reply]
Then set it lower still? That said...

> Because I test if there is a stack overflow error when testing TCO (tail call optimization)

You're making a shaky assumption there. Optimizing compilers (V8 is one) can prove all kinds of properties about your code.

If V8 can prove your tail call is not externally observable<sup>1</sup>, it may turn it into a loop or eliminate it entirely when it has no side effects. An example:
```js
let x=0;
function f(n) { if (x=n) f(n-1) }
f(42) // same as x=0, may be eliminated entirely
```

<sup>1</sup> through things like stack traces, although I believe V8 knows how to reconstruct stack traces for elided stack frames (it definitely knows how to do it for inlined functions) so even the potential presence of exceptions may not hamper it.

**Reply by jcubic** [reply]
My code is not tail recursive, it's not as trival as your example, it's Scheme interpreter in JavaScript and it was not optimized for TCO. The code was throwing the error, but now it's not when the recursive limit was increased. I call recursive factorial 
`(! 1000)` in my Scheme, and it overflows if I change it to `(! 2000)` but it will require me to update my tests when NodeJS or V8 decide to change the limit.
